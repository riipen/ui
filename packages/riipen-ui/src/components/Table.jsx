import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";
import withForwardedRef from "../utils/withForwardedRef";

class Table extends React.Component {
  static propTypes = {
    /**
     * What background to render for the table.
     * Defaults to grey100.
     */
    backgroundColor: PropTypes.oneOf([
      "transparent",
      "grey50",
      "grey100",
      "grey200",
      "grey300",
      "grey400"
    ]),

    /**
     * Children to render in the table element.
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element.
     */
    classes: PropTypes.arrayOf(PropTypes.string),

    /**
     * A reference to the table DOM element.
     */
    forwardedRef: PropTypes.element,

    /**
     * Whether to give all columns in the table equal width.
     */
    layout: PropTypes.oneOf(["auto", "fixed"])
  };

  static defaultProps = {
    backgroundColor: "grey100",
    layout: "auto"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      layout,
      backgroundColor,
      forwardedRef
    } = this.props;

    const theme = this.context;

    const tableBackground = {
      transparent: "transparent",
      grey50: theme.palette.grey[50],
      grey100: theme.palette.grey[100],
      grey200: theme.palette.grey[200],
      grey300: theme.palette.grey[300],
      grey400: theme.palette.grey[400]
    };

    return (
      <React.Fragment>
        <table ref={forwardedRef} className={clsx(classes)} border="0">
          {children}
        </table>
        <style jsx>{`
          table {
            background-color: ${tableBackground[backgroundColor]};
            border-radius: 4px;
            border-collapse: collapse;
            table-layout: ${layout};
            width: 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withForwardedRef(withClasses(Table));
