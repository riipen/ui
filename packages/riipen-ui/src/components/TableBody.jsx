import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

class TableBody extends React.Component {
  static propTypes = {
    /**
     * Children to render in the tbody element
     */
    children: PropTypes.node,

    /**
     * Classes to apply to the root element
     */
    classes: PropTypes.arrayOf(PropTypes.string)
  };

  render() {
    const { children, classes, ...other } = this.props;
    return (
      <tbody classes={clsx(classes)} {...other}>
        {children}
      </tbody>
    );
  }
}

export default withClasses(TableBody);
