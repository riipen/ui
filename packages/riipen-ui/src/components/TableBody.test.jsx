import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import TableBody from "./TableBody";
import Table from "./Table";

describe("<TableBody>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(
        <Table>
          <TableBody />
        </Table>
      );
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(
      <Table>
        <TableBody />
      </Table>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = (
        <tr>
          <td>Test</td>
        </tr>
      );

      const wrapper = mount(
        <Table>
          <TableBody>{child}</TableBody>
        </Table>
      );

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the tbody node", () => {
      const classes = ["classOne"];

      const wrapper = mount(
        <Table>
          <TableBody classes={classes} />
        </Table>
      );

      expect(wrapper.find("tbody").props().classes).toContain(classes[0]);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-tablebody"];

      const wrapper = mount(
        <Table>
          <TableBody />
        </Table>
      );

      expect(
        wrapper
          .find("TableBody")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });
});
