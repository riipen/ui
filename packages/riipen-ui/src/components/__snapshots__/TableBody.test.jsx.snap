// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableBody> renders correct snapshot 1`] = `
<ForwardRef(withForwardedRef(withClasses(Table)))>
  <withClasses(Table)
    forwardedRef={null}
  >
    <Table
      backgroundColor="grey50"
      classes={
        Array [
          "riipen",
          "riipen-table",
        ]
      }
      forwardedRef={null}
      layout="auto"
    >
      <table
        border="0"
        className="jsx-1059775881 riipen riipen-table"
      >
        <withClasses(TableBody)>
          <TableBody
            classes={
              Array [
                "riipen",
                "riipen-tablebody",
              ]
            }
          >
            <tbody
              classes="riipen riipen-tablebody"
            />
          </TableBody>
        </withClasses(TableBody)>
      </table>
      <JSXStyle
        dynamic={
          Array [
            "#fafafa",
            "auto",
          ]
        }
        id="4123435624"
      />
    </Table>
  </withClasses(Table)>
</ForwardRef(withForwardedRef(withClasses(Table)))>
`;
